wallet-new/src/Cardano/Wallet/Kernel/Restore.hs

```diff
@@ -379,54 +381,50 @@ restoreWalletHistoryAsync wallet rootId prefilter progress start (tgtHash, tgtSl
     startingPoint <- case start of
         Nothing -> withNode $ getFirstGenesisBlockHash genesisHash
         Just sh -> nextHistoricalHash sh >>= maybe (throwM $ RestorationSuccessorNotFound sh) pure
-    restore genesisHash startingPoint NoTimingData
+    restore genesisHash startingPoint 1000
   where
     wId :: WalletId
     wId = WalletIdHdRnd rootId
 
     -- Process the restoration of the block with the given 'HeaderHash'.
-    restore :: GenesisHash -> HeaderHash -> TimingData -> IO ()
-    restore genesisHash hh timing = do
-
-        -- Updating the average rate every 5 blocks.
-        (rate, timing') <- tickTiming 5 timing
-
-        -- Update each account's historical checkpoints
-        block <- getBlockOrThrow genesisHash hh
+    restore :: GenesisHash -> HeaderHash -> Int -> IO ()
+    restore genesisHash hh batchSize = do
+        -- Keep track of timing
+        startTime <- getCurrentTime
+
+        -- Processing a batch of blocks, flushing to Acid-State only once
+        (updates, (hh', finished)) <- flip runStrictStateT (hh, False) $ sequence $ replicate batchSize $ do
+            (currentHash, finished) <- get
+            if finished then return Nothing else do
+                if tgtHash == currentHash then
+                    put (currentHash, True)
+                else
+                    lift (nextHistoricalHash currentHash) >>= \case
+                        Nothing  -> throwM (RestorationFinishUnreachable tgtHash currentHash)
+                        Just hh' -> put (hh', False)
+                lift $ getPrefilteredBlockOrThrow genesisHash currentHash
```

Comment:
The use of a state monad here seems a bit odd to me. This is just keeping the `currentHash` as the running value right? I think I'd just write this as a local recursive action with the `currentHash` as a local arg.